pkg/privilege/privileges/cache.go

```diff
@@ -400,6 +410,121 @@ func (p *MySQLPrivilege) LoadAll(ctx sessionctx.Context) error {
+	}
+
+	p.RoleGraph = make(map[string]roleGraphEdgesTable)
+	err = p.loadTable(ctx, sqlLoadRoleGraph, p.decodeRoleEdgesTable)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+func dedupSorted[S ~[]E, E any](s S, eq func(a, b E) bool) S {
+	skip := 0
+	for i := 1; i < len(s); i++ {
+		if eq(s[i], s[i-1]) {
+			skip++
+		}
+		s[i-skip] = s[i]
+	}
+	s = s[:len(s)-skip]
+	return s
+}
+
+// merge construct a new MySQLPrivilege by merging the data of the two objects;.
+func (p *MySQLPrivilege) merge(diff *immutable) *MySQLPrivilege {
+	var ret MySQLPrivilege
+	ret.User = make([]UserRecord, 0, len(p.User)+len(diff.User))
+	ret.User = append(ret.User, p.User...)
+	ret.User = append(ret.User, diff.User...)
+
+	// sort and dedup
+	slices.SortStableFunc(ret.User, compareUserRecord)
+	ret.User = dedupSorted(ret.User, func(x, y UserRecord) bool { return x.User == y.User && x.Host == y.Host })
+	ret.buildUserMap()
+
+	ret.DB = make([]dbRecord, 0, len(p.DB)+len(diff.DB))
+	ret.DB = append(ret.DB, p.DB...)
+	ret.DB = append(ret.DB, diff.DB...)
+	ret.buildDBMap()
+
+	ret.TablesPriv = make([]tablesPrivRecord, 0, len(p.TablesPriv)+len(diff.TablesPriv))
+	ret.TablesPriv = append(ret.TablesPriv, p.TablesPriv...)
+	ret.TablesPriv = append(ret.TablesPriv, diff.TablesPriv...)
+	ret.buildTablesPrivMap()
+
+	ret.ColumnsPriv = make([]columnsPrivRecord, 0, len(p.ColumnsPriv)+len(diff.ColumnsPriv))
+	ret.ColumnsPriv = append(ret.ColumnsPriv, p.ColumnsPriv...)
+	ret.ColumnsPriv = append(ret.ColumnsPriv, diff.ColumnsPriv...)
+	slices.SortStableFunc(ret.ColumnsPriv, compareColumnsPrivRecord)
+	ret.ColumnsPriv = dedupSorted(ret.ColumnsPriv, func(x, y columnsPrivRecord) bool {
+		return x.Host == y.Host && x.User == y.User &&
+			x.DB == y.DB && x.TableName == y.TableName && x.ColumnName == y.ColumnName
+	})
+
+	ret.DefaultRoles = make([]defaultRoleRecord, 0, len(p.DefaultRoles)+len(diff.DefaultRoles))
+	ret.DefaultRoles = append(ret.DefaultRoles, p.DefaultRoles...)
+	ret.DefaultRoles = append(ret.DefaultRoles, diff.DefaultRoles...)
+	slices.SortStableFunc(ret.DefaultRoles, compareDefaultRoleRecord)
+	ret.DefaultRoles = dedupSorted(ret.DefaultRoles, func(x, y defaultRoleRecord) bool {
+		return x.Host == y.Host && x.User == y.User
+	})
+
+	ret.DynamicPriv = make([]dynamicPrivRecord, 0, len(p.DynamicPriv)+len(diff.DynamicPriv))
+	ret.DynamicPriv = append(ret.DynamicPriv, p.DynamicPriv...)
+	ret.DynamicPriv = append(ret.DynamicPriv, diff.DynamicPriv...)
+	ret.buildDynamicMap()
+
+	ret.GlobalPriv = make([]globalPrivRecord, 0, len(p.GlobalPriv)+len(diff.GlobalPriv))
+	ret.GlobalPriv = append(ret.GlobalPriv, p.GlobalPriv...)
+	ret.GlobalPriv = append(ret.GlobalPriv, diff.GlobalPriv...)
+	slices.SortStableFunc(ret.GlobalPriv, compareGlobalPrivRecord)
+	ret.GlobalPriv = dedupSorted(ret.GlobalPriv, func(x, y globalPrivRecord) bool {
+		return x.Host == y.Host && x.User == y.User
+	})
+	ret.buildGlobalMap()
+
+	ret.RoleGraph = diff.RoleGraph
```

Comment:
RoleGraph is a full load, because I found that the table definitions differ from the others.
Other other tables has the User+Host column, while this table does not have it.